### DISCUSSION TITLE
Why can't I fix the sampler results?

### DISCUSSION THREAD

**Discussion by idssplab** [optuna/optuna#5219] [Q&A] [answered]
Thank you for checking this question🙏
I'm not sure whether I understand the random seed well or not.

When I used PyTorch Lightning and Optuna to do Hyper parameter training,
I want to fix the training results.

I tried to set random seed through this way:
```
import torch
import numpy as np
import random
import os
from lightning.pytorch import seed_everything

random_seed = 42

# 設置隨機種子
def set_seed(seed=42):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)  # if you are using multi-GPU.
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False  

# 設置一個固定的種子
set_seed(random_seed)

seed_everything(random_seed, workers=True)
```

And this is my objective function:
```
# Objective 函數
def objective(trial):
    # 定義超參數搜索空間
    hidden_layer_size = trial.suggest_int('hidden_layer_size', 10, 100)
    lr = trial.suggest_loguniform('lr', 1e-5, 1e-1)
    max_epochs = trial.suggest_int('max_epochs', 10, 50)
    batch_size = trial.suggest_categorical('batch_size', [1, 2, 4, 8, 16, 32])

    # 進行交叉驗證
    auprc_scores_cv = []
    kfold = KFold(n_splits=5, shuffle=True, random_state=42)
    for fold, (train_ids, test_ids) in enumerate(kfold.split(train_dataset)):
        # 創建數據加載器
        train_subset = Subset(train_dataset, train_ids)
        test_subset = Subset(train_dataset, test_ids)
        train_loader = DataLoader(train_subset, batch_size=batch_size, shuffle=True)
        test_loader = DataLoader(test_subset, batch_size=1, shuffle=False)

        # 初始化模型並設置超參數
        input_size = sequences_tensor.shape[2] * 7
        model = RNNModel(input_size, hidden_layer_size=hidden_layer_size)
        model.configure_optimizers = lambda: torch.optim.Adam(model.parameters(), lr=lr)

        # 訓練模型
        trainer = pl.Trainer(max_epochs=max_epochs, deterministic=True)
        trainer.fit(model, train_loader)
        
        try:
            # 計算指標
            _, auprc, _, _, _, _, _ = calculate_metrics(model, test_loader)
            auprc_scores_cv.append(auprc)
        except ValueError as e:
            if "Only one class present in y_true" in str(e):
                # 當捕捉到特定錯誤時，返回一個預設值
                auprc = 0.00  # 默認值，可以根據你的需求調整
            else:
                raise  # 如果是其他類型的 ValueError，重新拋出錯誤

    # 計算平均 AUPRC
    average_auprc = sum(auprc_scores_cv) / len(auprc_scores_cv)
    return average_auprc 


# 創建一個 Optuna Study
study = optuna.create_study(direction='maximize', sampler=optuna.samplers.TPESampler(seed=random_seed))

# 執行優化
study.optimize(objective, n_trials=50)
```

I originally thought that I set the random seed, set the Trainer's deterministic True,
and set the sampler's seed so the results could be fixed.
But that way still made different results every time.

**Comment by nzw0301** [comment]
Could you share minimal reproducible code with us? The current code does not define a dataset. Dummy data would be enough.
In addition, version info of optuna, pytorch, and pytorch-lightning is helpful.

**Reply by idssplab** [reply]
Thanks for your response 🙏
I will prepare these information as soon as possible.

**Reply by nzw0301** [reply]
Lovely. Thanks!

**Reply by idssplab** [reply, marked as answer]
Thanks again!
But we finally decided to preserve the hyper parameter training result,
and then using that result to reproduce our experiments!

If we still wonder this problem, 
we will prepare those information and start a discussion again.
Thank you for your time!

**Reply by nzw0301** [reply]
Thank you for your update. Glad to hear tha! I think it is not necessary since it takes time to do so.
